keps/sig-release/20190316-rebase-images-to-distroless.md

```diff
@@ -0,0 +1,149 @@
+### Distroless and Previous Work
+"Distroless images contain only your application and its runtime dependencies. They do not contain package managers, shells or any other programs you would expect to find in a standard Linux distribution.”  (from [github.com/GoogleContainerTools/distroless/README](https://github.com/GoogleContainerTools/distroless))
+Distroless supports the dependencies where “FROM scratch” misses and more light-weighted than debian or alpine. Meanwhile, distroless “improves the signal to noise of scanners (e.g. CVE) and reduces the burden of establishing provenance to just what you need.”(from [github.com/GoogleContainerTools/distroless/README](https://github.com/GoogleContainerTools/distroless))
+
+Using Distroless as a common image base is originally proposed as an exploration area in [the base image policy for system containers](https://github.com/kubernetes/kubernetes/issues/40248). Tim(tallclair@) has driven the effort on defining and establishing the base image policy (main changes):
+
+* Add Alpine iptable as base image for kube-proxy. Previously kube-proxy is based on debian iptable image.  (This direction is scraped, see [issue/39696](https://github.com/kubernetes/kubernetes/issues/39696) for details)
+Rebase busybox images to debian-base.
+* Rebase busybox images to debian-base.
+* Rebase certain alpine images to debian-base
+* Image related works on migrating apiserver-builder to kube-builder
+
+The distroless solution is filed separately in [issue/70249](https://github.com/kubernetes/kubernetes/issues/70249). The doc, as a more up-to-date version, is slightly different than the original issue.
+
+## Proposal
+
+The approaches to rebase different containers vary significantly due to the technical implementations, development plans or even the legacy reasons. For example, [ingress-gce/fuzzer](https://github.com/kubernetes/ingress-gce/blob/64eee7e3521680057b071d5e9bebaa215086a4bc/Dockerfile.fuzzer) was based on alpine and can't be switched to distroless directly due to the fact that it needs shell to redirect the glog file. To allow the images to be based on distroless (which doesn't contain shell), we firstly need to remove the dependency on the shell (use klog instead of glog), and then rebase the image. (Related PR [pr/682](https://github.com/kubernetes/ingress-gce/pull/682]), [pr/666](https://github.com/kubernetes/ingress-gce/pull/666))
+There are 4 main efforts on this KEP.
+
+1. The images are built via bazel. In such case, we will update the bazel BUILD rule to switch to base from distroless image. This method applies for the core containers like kube-apiserver, kube-controller-manager, cloud-controller-manager, kube-scheduler. As for kube-proxy, it is based on debian-iptable which distroless doesn't support iptables yet. Thus, kube-proxy won't be changed.
```

Comment:
What is the level of effort needed to get iptables installed in a kube-proxy image based on distroless? Seems like if we do not convert all the standard shipping images, the benefit of reduced CVE tracking, etc. is largely lost.

Reply:
Agree. This KEP is targeting 1.15 so I only put what can guarantee to be done here. 

For kube-proxy, I think the main blocker is ["kube-proxy writes the iptables rules which implement Service". ](https://kubernetes.io/docs/tasks/debug-application-cluster/debug-service/#is-kube-proxy-writing-iptables-rules). So two approaches can be done: one is changing the way how Service is implemented and the other is requesting distroless to support iptable. Either sounds like significant changes and deserve further in-depth discussions.

@tallclair  Do you have any insights on this? It will be great if distroless can support iptable.

Reply:
I am wondering if it is feasible to build static iptables binaries and include them, not in distroless, but in the kube-proxy image that is based on distroless.

Reply:
Couldn't help myself and I poked around at this some today. Getting statically built iptables is a pain, but I was able to get iptables built and working using gcr.io/distroless/cc:latest:

```
FROM gcc:latest
RUN curl https://www.netfilter.org/projects/iptables/files/iptables-1.8.2.tar.bz2 --output - \
  | bunzip2 -c | tar xv \
  && mkdir /built \
  && cd iptables-1.8.2 \
  && ./configure --prefix /built --enable-static --disable-nftables --disable-shared \
  && make && make install

FROM gcr.io/distroless/cc:latest
COPY --from=0 /built/sbin/* /sbin/
```

Reply:
My concern is that we lose any insights into iptables CVEs if we do this. We also don't get to benefit from debian backporting patches for those CVEs.

As an alternative solution, what if we just took the original `FROM debian-base` image, and then copied iptables and the linked libraries [1] onto distroless/static?

I think it's worth exploring these options, but I also think we should keep it out of scope for this KEP so we can focus on the low hanging fruit. Even if we don't migrate every image off debian base, it will still be a substantially reduced maintenance burden for us (speaking as someone who is primarily responsible for rebuilding these images when a CVE is found).

[1]
```
ldd /sbin/iptables                          
	linux-vdso.so.1 (0x00007fff727f3000)
	libip4tc.so.0 => /usr/lib/x86_64-linux-gnu/libip4tc.so.0 (0x00007fa131a6f000)
	libip6tc.so.0 => /usr/lib/x86_64-linux-gnu/libip6tc.so.0 (0x00007fa131867000)
	libxtables.so.12 => /usr/lib/x86_64-linux-gnu/libxtables.so.12 (0x00007fa13165a000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fa131356000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fa130fb7000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fa130db3000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fa131ef4000)
```

Reply:
The debian-iptables image includes several dependencies, not just iptables, though:
https://github.com/kubernetes/kubernetes/blob/e6272b887b81a62e6f06b7fac4b3b61d1c8bf657/build/debian-iptables/Dockerfile#L18-L23